src/axes.jl

```diff
@@ -248,3 +247,64 @@ lazy_axes(x::AbstractRange, ::StaticInt{1}) = Base.axes1(x)
 lazy_axes(x, ::Colon) = LazyAxis{:}(x)
 lazy_axes(x, ::StaticInt{dim}) where {dim} = ndims(x) < dim ? SOneTo{1}() : LazyAxis{dim}(x)
 @inline lazy_axes(x, dims::Tuple) = map(Base.Fix1(lazy_axes, x), dims)
+
+"""
+    index_labels(x)
+    index_labels(x, dim)
+
+Returns a tuple of labels assigned to each axis or a collection of labels corresponding to
+axis `dim` of `x`. Default is to simply return `map(keys, axes(x))`.
```

Comment:
Is it really a good idea to have the default return `keys`? I'd say better return `nothing` so that the caller knows that no labels are set (it's easy to call `keys` if needed). Otherwise there's no way to know whether labels are present or not.

Reply:
I should probably have changed that back to just the axis after all the discussion around naming this. My thinking was that it would be odd if we returned something that didn't "label" the indexes at all because we are assuming that is what's returned.

Returning `nothing` would require that every method that acts on index labels would require and extra step to check if it can actually look up values in it. Not an insurmountable issue, but it would certainly be tedious.

Throwing an error when this isn't defined also isn't an option because things like a vector transpose will inevitably have a dimension without labels.

Reply:
If we don't return `nothing` when no labels are set, then we should add a separate function called e.g. `has_index_labels`, a bit like we did in the Metadata PR with `hasmetadata` and `metadata`.

Reply:
That's definitely doable and is probably a good method to have irregardless of what the default we choose.

I'm not against returning something that clearly shows no labels are defined. Perhaps I'm being too strict with the rule that the return has the same indices as their corresponding axis, even for default values. I assume people will at least expect to be able to do things like `findfirst(==(label), index_labels(A, dim))`, in which case we would probably still want a more informative error.

Reply:
I would also prefer `nothing`

Reply:
I think the issues I mentioned can be avoided with documentation and tooling if we decide to return nothing. But AxisKeys.jl uses a similar approach to what's implemented here so it would probably be good if @mcabbott let us know if we can make this work for AxisKeys

Reply:
The issue for DimensionalData.jl is to be able to not propagate the index labels when there are none, such as in`cat`, where joining multiple `Base.OneTo` ranges from `keys` is really not what you want to do.

In DD I use an explicit `NoLookup` object to solve this - it works correctly in both `findfirst` and `cat`. The problem for me will be interop with other types, where there needs to be a way of knowing if there are real index labels associated with an axis, not just `keys`.

Probably a `has_index_labels` trait would work for this too, maybe it is better than returning `nothing`.

Reply:
Good point. A default without a clearly know type may cause more problems than it solves

Reply:
> In DD I use an explicit NoLookup object to solve this - it works correctly in both findfirst and cat. The problem for me will be interop with other array types, where there needs to be a way of knowing if there are real index labels associated with an axis, not just keys.

What about returning an `UndefLabels`?

Reply:
> Is it really worth defining a special type just for that? What would be the advantage over Nothing?

I'm just trying to figure out some generic representation of labels not being defined that can pass between methods. If we do `vcat(a, b)` but one doesn't have index labels, then we can't do `vcat(index_labels(a, 1), index_labels(b, 1))`. Even if we define something that doesn't error in this situation (e.g., `vcat_index_labels(::AbstractVector, ::Nothing)`), how does this produce a new vector of index labels for the newly generated array? I figured `UndefLabel` could represent every index without a label.

@rafaqz how are you handling this situation?

Reply:
@rafaqz, it looks like `NoLookup` is basically the same thing as what I'm describing. It wraps the non-labeled indices.

Reply:
Also, do we want to have `has_index_labels(T::Type)` or `has_index_labels(x) = any(are_labels, index_labels)`?

Reply:
Sorry I missed this earlier. If your `UndefLabels` is essentially my `NoLookup`, that should be fine.